### ISSUE TITLE
Adjacency and Incidence matrix of node tree / directed graph

### ISSUE THREAD

**Issue by zeffii** [zeffii/SoundPetal#5]
protocode

Adjacency Matrix

``` python
import bpy
import numpy as np

ng = bpy.data.node_groups[0]

num_nodes = len(ng.nodes)
nodes = ng.nodes

node_table = {node.name: idx for idx, node in enumerate(nodes)}

M = np.zeros((num_nodes, num_nodes))
for link in ng.links:
    x = node_table.get(link.from_node.name)
    y = node_table.get(link.to_node.name)
    M[x][y] += 1

print(M)
```

**Comment by ly29** [comment]
An interesting approach.

**Comment by zeffii** [comment]
@ly29 Flow needs a working dependency graph before I go any further.

``` python
# adjacency [M] , incidence matrix [ I ]

import bpy
import numpy as np

ng = bpy.data.node_groups[0]

num_nodes = len(ng.nodes)
num_links = len(ng.links)
nodes = ng.nodes

node_table = {node.name: idx for idx, node in enumerate(nodes)}

M = np.zeros((num_nodes, num_nodes))
I = np.zeros((num_nodes, num_links))
for idx, link in enumerate(ng.links):
    name_x = link.from_node.name
    name_y = link.to_node.name
    x = node_table.get(name_x)
    y = node_table.get(name_y)
    M[x][y] += 1
    I[x][idx] += 1    
    I[y][idx] += 1    

print(M)
print(I)
```

**Comment by ly29** [comment]
For Sverchok I did it easier and just captured the sorted output of the dependency graph.
These are available for separated node sets in a layout. If I redid today I might even do a recursive variant since I now know that the slow step was to parse it node for node, socket for socket
The main function is `make_update_list` but before that I prepare the data. It could be quicker perhaps but it is okay.

``` python
        node_sets = separate_nodes(ng)
        deps = make_dep_dict(ng)
        out = [make_update_list(ng, s, deps) for s in node_sets]
```

https://github.com/nortikin/sverchok/blob/toProcess/core/update_system.py#L71-L124

**Comment by ly29** [comment]
Also something to consider is to create a whole separated structure to sort away syntactic sugar like reroutes and similar things (like wifi nodes) if you decide to add them.

**Comment by zeffii** [comment]
Thanks @ly29 ! i haven't taken a serious stab at the update mechanism yet, but what's there works for now. The problem really pops up when I hit F8 this is what I want to focus on first. that and the registration of new nodes between F8.

**Comment by ly29** [comment]
I spent a lot of time on the F8 reload in Sverchok.
I the end I was only able to tag the events and launching a full update event when somebody triggers anything in Sverchok.

```
__init__.py
```

``` python
if "bpy" in locals():
   # a lot of things
   reload_event = True

def register():
    if reload_event:
        # register all
        for m in imported_modules:
            if m.__name__ == "data_structure":
                m.RELOAD_EVENT = True
```

And then this is checked in the update system. Didn't find a way to add it otherwise, because the register context is limited and many things don't work then. I think the `post_load_handler` should be called after F8, but no...

``` python
def process_tree(ng=None):
    global update_cache
    global partial_update_cache

    if data_structure.RELOAD_EVENT:
        reload_sverchok()
        return
```

Anyway a blender startup takes 5 s for me so I don't care that much when I need to restart it.

**Comment by ly29** [comment]
New nodes works in sverchok. `importlib.reload` needs to be called on all modules before acting on them in a case of reload event I think.

``` python
if "bpy" in locals():
    import nodeitems_utils
    nodes = importlib.reload(nodes)
    node_list = make_node_list()
    for im in imported_modules+node_list:
        importlib.reload(im)
```

**Comment by zeffii** [comment]
Cool. New nodes are now detected. \o/

**Comment by zeffii** [comment]
It's easy to find 
- all nodes that are subgraphs (disjoint) (also no in-out )
- all nodes that are the start of a chain

What i'm trying to solve now (via matrix manipulation) is what order to trigger updates. I know this is a solved problem in Sverchok, this project is purely for intellectual stimulation

**Comment by ly29** [comment]
That is also quite easy. But I will leave you to do it.

**Comment by zeffii** [comment]
I don't see that answer so fast, but that doesn't bother me. Plenty of other smaller things to fix for `0.2`
